keps/sig-storage/1472-storage-capacity-tracking/README.md

```diff
@@ -767,17 +830,25 @@ enhancement:
       - kube-scheduler
 
 * **Does enabling the feature change any default behavior?**
-  No changes for existing storage driver deployments.
+
+  Enabling it only in kube-scheduler and api-server and not any of the
+  running CSI drivers causes no changes. Everything continues as
+  before because no `CSIStorageCapacity` objects are created and
+  kube-scheduler does not wait for any.
+
+  That changes once the feature is enabled in a CSI driver. Then pod
+  scheduling becomes more likely to pick suitable nodes. This happens
+  automatically, without having to change application deployments.
 
 * **Can the feature be disabled once it has been enabled (i.e. can we rollback
   the enablement)?**
   Yes.
 
-  When disabling support in the API server, the new object type and
-  thus all objects will disappear together with the new flag in
+  When disabling support in the API server, the new object type
```

Comment:
We need to distringuish two things:
- version rollback - in which case the type will no longer be server (but please be explicit it's about version rollback)
- feature disablemenet (via feature gate) - in this case the type will still be visible

Reply:
Perhaps it got lost in the GitHub noise, but I replied to that already here: https://github.com/kubernetes/enhancements/pull/1829/files/b039a48bffcab54f98975f4dfdb2df7e0f9944ad#r566714065

I said:

> Depending on how thoroughly the feature gets disabled, the API is not there anymore. It will be gone when any of the following happens for the apiserver:
> - the v1alpha1 API group gets disabled (not relevant anymore once it is beta)
> - the feature is turned off for the apiserver
> - the apiserver is rolled back to a version without the API
> 
> In this section, I already tried to discuss disabling the feature in the three places where it could be disabled separately (apiserver, scheduler, CSI driver). I'm not sure what to add?

Just to be very clear, in the second case that you mentioned ("feature disablement (via feature gate)") the type will *also* be gone and not visible. So there is really no difference between that case and version rollback.

Reply:
> Just to be very clear, in the second case that you mentioned ("feature disablement (via feature gate)") the type will also be gone and not visible. 

It shouldn't - we generally don't hide type registration behind feature gate.

Reply:
Somehow this is common for storage types. For example: https://github.com/kubernetes/kubernetes/blob/73dd5c840662bb066a146d0871216333181f4b64/pkg/registry/storage/rest/storage_storage.go#L98-L105

@wojtek-t and I were discussing this on chat and weren't sure whether this is a good or bad thing. He wanted to clarify with Jordan.

In the meantime I have changed the first sentence to make it clear that registration is currently depending on the feature gate.

Reply:
I just briefly chatted with Jordan about this offline.

Given that APIs are controlled via `--runtime-config` flag anyway, then having that controlled via feature gate doesn't bring any value. So the consensus we get to is that you basically should remove that feature gate from there.

[And then it will evolve to situation I was suggesting above.]

Please also add to the KEP somewhere that feature-gate will be removed from controlling API registration.

Reply:
> Given that APIs are controlled via --runtime-config flag anyway, then having that controlled via feature gate doesn't bring any value. So the consensus we get to is that you basically should remove that feature gate from there.

That argument currently applies because it is in the v1alpha1 group. But once we move it to v1beta1 in 1.20, it will become impossible to remove just that one type.

I can see pros and cons for both approaches: when the apiserver can disable the type, the cluster can be recovered from a malfunctioning CSI driver quickly by disabling the type and later fixing the CSI driver. The downside of disabling the type is that it is impossible to delete stale objects.

Overall I prefer to not have the feature gate check.

> Please also add to the KEP somewhere that feature-gate will be removed from controlling API registration.

Done.